### ISSUE TITLE
Relays can cause delays to clients and not be penalized

### ISSUE THREAD

**Issue by nventuro** [tabookey/tabookey-gasless#160]
From the EIP:
>Attack: Relay attempts to censor a transaction by signing it, but publishing a different 
transaction with the same nonce.
>
>Reusing the nonce is the only DoS performed by a Relay, that cannot be detected within a couple of seconds during the http request. It will only be detected when the malicious transaction with the same nonce gets mined and triggers the RelayHub.TransactionRelayed event.

This attack is mitigated by the penalization mechanism, which will let any account penalize the relay, causing it to lose half of its stake. However, this is the actual implementation in the contract:

```
function penalizeRepeatedNonce(bytes memory unsignedTx1, bytes memory signature1, bytes memory unsignedTx2, bytes memory signature2) public {
    address addr1 = keccak256(abi.encodePacked(unsignedTx1)).recover(signature1);
    address addr2 = keccak256(abi.encodePacked(unsignedTx2)).recover(signature2);

    require(addr1 == addr2, "Different signer");

    Transaction memory decodedTx1 = decodeTransaction(unsignedTx1);
    Transaction memory decodedTx2 = decodeTransaction(unsignedTx2);

    require(decodedTx1.nonce == decodedTx2.nonce, "Different nonce");
    require(keccak256(abi.encodePacked(decodedTx1.data)) != keccak256(abi.encodePacked(decodedTx2.data)), "tx.data is equal");

    penalize(addr1);
}
```

Two transactions with the same nonce are not enough to cause a penalization: the data field must be different. This achieves two purposes:
 - a transaction cannot be penalized against itself
 - a relay may resend a previously sent transaction using a higher gas price

However, this opens the door to relay attacks. A relay may also:
 - reuse the nonce, with the same data, and _a lower gas limit_, causing the `relayCall` call to not succeed due to not enough gas
 - reuse the nonce, with the same data, and _give value to it_, causing the `relayCall` call to revert due to the function not being `payable`

In both scenarios, the relay caused a delay on the client, but it cannot be penalized by the current mechanisms. We need to update the `penalizeRepeatedNonce` so that these cases are handled.

**Comment by drortirosh** [comment]
Legitimate relayed requests will always have a different data on each call: even if comes from the same client, each transaction carries different nonce.
So if a relay respond to the client with wrong transaction (that is, its must be the exact same content, signed, with the only allowed difference of higher gas-price), the client can simply go to a different relay (and locally save a bad reputation for this one)
This effectively achieves a delay of one network roundtrip: since the client knows the relay is alive (it answered a "ping"), it can way ~10 seconds and go to the next relay

The worst thing (to the client) the relay can do then is actually submit the ORIGINAL transaction on-chain: the effect is cause the client to appear as "spamming": the client had sent 2 transactions to 2 different relays.

In isolation, this is not a problem: the client won't use this relay anymore, and one on-chain "replay" is not considered bad reputation for the client.

**Comment by nventuro** [comment]
The scenario I was trying to describe is not the relay replying to the user with an incorrect transaction, but rather, replying with the correct one, _but broadcasting a different one_ (see the bit of the EIP that I quoted on the original post).

Example:
 1. User requests transaction with parameters a, b, c and d
 2. Relay signs transaction with parameters a, b c, d and nonce n, sends it back to the user
 3. Relay immediately broadcasts _another_ transaction with nonce n, but this time a) using a lower gas limit, or b) adding value to it (i.e. `msg.value > 0`)
 4. Only one of the transactions in 2. and 3. will be included in a block. If the one in 3. is included, it will revert due to invalid parameters, and the user won't notice it until the block is finalized (delay)
 5. User cannot penalize the relay, since changing value or gas limit is not considered penalizable behavior.

**Comment by shahafn** [comment]
Just fixed the issue, openning a PR soon.

**Comment by shahafn** [comment]
Fixed in #176

**Comment by nventuro** [comment]
>gas limit should not change between txs and should be penalizable, otherwise the relay can deny service of the user by sending a much lower gas limit tx onchain giving the client higher gas limit tx. The onchain tx will get reverted, and the client will only figure that out after it was mined.

@shahafn yes, this is the scenario I described above. Perhaps my original message wasn't clear?

We can close this now that #176 is merged.
